src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php

```diff
@@ -296,6 +296,11 @@ private function registerBusMiddleware(ContainerBuilder $container, string $busI
                 $childDefinition = new ChildDefinition($messengerMiddlewareId);
                 $count = \count($definition->getArguments());
                 foreach (array_values($arguments ?? array()) as $key => $argument) {
+                    // Convert argument to references if needed.
+                    if (is_string($argument) && '@' === $argument[0]) {
+                        $argument = new Reference(substr($argument, 1));
```

Comment:
> Ping @ogizanagi, This is the best way I found to support references. Im not sure if this is needed if you configure framework.messenger with XML.

Or should I go the long way of creating a Factory that uses `@container` as a dependency?

Reply:
The idea of factories weren't to replace DI config when more suitable, only to solve most simple needs for configurable middlewares. Therefore, I don't think we should add services references support.
If really desired, we could indeed use dedicated service locators or perhaps let the user wiring it itself by creating the middleware service using DI config for now.

Reply:
Hm. I tried using a Factory that depends on `@service_container`. That does not work since the buses are private. 

I also tried injecting all services tagged with `message_bus` into the factory. That did not work because of circular dependencies.

>  Circular reference detected for service "debug.traced.messenger.bus.command", path: "debug.traced.messenger.bus.command -  
  > debug.traced.messenger.bus.command.inner -> messenger.bus.command.middleware.messenger.middleware.handles_recorded_mess  
  ages -> messenger.middleware.handles_recorded_messages.factory -> debug.traced.messenger.bus.command". 

Im saying that I do not know a better (working) solution that this. Im sorry.

Reply:
Injecting the Symfony service container now belong to another era 😃 
We should register dedicated [service locators](https://symfony.com/blog/new-in-symfony-3-3-service-locators) instead. This would happen in the `MessengerPass`, using `ServiceLocatorTagPass::register()`.

But as I said, we could let the user create the concrete instance of this middleware in userland for now.

Reply:
Done. I've updated the configuration example